GAD 9 - AVL Baum/tests/avl/AVLTreeTester.java

```diff
@@ -0,0 +1,316 @@
+        root.setBalance(-1);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+
+        try {
+            a.setRight(root);
+            a.setBalance(1);
+            assertFalse(tree.validAVL(), "a tree must not contain circles");
+            a.setRight(null);
+            a.setBalance(0);
+            assertTrue(tree.validAVL(), "correct AVL tree");
+        } catch (StackOverflowError exc) {
+            fail("StackOverflow: fix problems with circular trees");
+        }
+
+        root.setLeft(b);
+        assertFalse(tree.validAVL(), "left key was bigger than root key");
+
+        root.setLeft(a);
+        root.setRight(b);
+        root.setBalance(0);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+
+        a.setLeft(a1);
+        a.setRight(a2);
+        b.setLeft(b1);
+        b.setRight(b2);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+
+        b.setRight(b1);
+        b.setLeft(b2);
+        assertFalse(tree.validAVL(), "node B children's keys have right < left");
+
+        b.setLeft(new AVLTreeNode(-3));
+        b.setRight(b2);
+        assertFalse(tree.validAVL(), "all nodes of all subtrees need to be sorted");
+
+        b.setLeft(b1);
+        b.setRight(null);
+        assertFalse(tree.validAVL(), "wrong balance at node B");
+
+        b.setBalance(-1);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+
+        root.setBalance(0);
+        a.setBalance(1);
+        assertFalse(tree.validAVL(), "wrong balance at node A");
+
+        a.setLeft(null);
+        assertTrue(tree.validAVL(), "correct AVL tree");
+    }
+
+    /**
+     * Count the number of nodes in the subtree, including the root itself.
+     *
+     * @param subtree root of a subtree
+     * @return number of nodes in the subtree
+     */
+    private int count(AVLTreeNode subtree) {
+        int l = subtree.getLeft() == null ? 0 : count(subtree.getLeft());
+        int r = subtree.getRight() == null ? 0 : count(subtree.getRight());
+        return 1 + l + r;
+    }
+
+    /**
+     * Test to check that insert() works to insert new values.
+     * No checks on the layout of the AVL tree are performed in
+     * this test. It will only check that the number of nodes
+     * in the tree is always correct after every insert() call.
+     */
+    @Test
+    public void testInsertNumberOfValues() {
+        AVLTree tree = new AVLTree();
+        tree.insert(42);
+        try {
+            assertEquals(42, tree.getRoot().getKey(), "root node has wrong value");
+        } catch (NullPointerException exc) {
+            fail("you need to implement insert() first");
+        }
+
+        assertEquals(1, count(tree.getRoot()), "expected one element in the tree");
+        Random random = new Random();
+        for (int i = 0; i < 999; i++) {
+            tree.insert(random.nextInt());
+            assertEquals(2 + i, count(tree.getRoot()), "expected " + (2 + i) + " nodes in the tree");
+        }
```

Comment:
You could test if the testees own `validAVL()` method accepts the tree.

Reply:
I've added such a check. But the check depends on a valid `validAVL` method, otherwise the check would be useless (as explained in the test case message).

Reply:
actually I meant within the loop on every insert ^^

Reply:
Ah, I see. Nevermind :)